## 1.0 IMPORT MODULES

In [7]:
# import numpy and pandas libraries
import numpy as np
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn import preprocessing
from sklearn.impute import SimpleImputer
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
from sklearn.tree import DecisionTreeClassifier 
from sklearn.metrics import confusion_matrix
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression

np.random.seed(1)

## 2.0 LOAD DATA

In [8]:
# load data
UniversalBank = pd.read_csv("UniversalBank.csv")

## 3.0 INITIAL EXPLORATION OF DATA

In [9]:
UniversalBank.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  5000 non-null   int64  
 1   Age                 5000 non-null   int64  
 2   Experience          5000 non-null   int64  
 3   Income              5000 non-null   int64  
 4   ZIP Code            5000 non-null   int64  
 5   Family              5000 non-null   int64  
 6   CCAvg               5000 non-null   float64
 7   Education           5000 non-null   int64  
 8   Mortgage            5000 non-null   int64  
 9   Personal Loan       5000 non-null   int64  
 10  Securities Account  5000 non-null   int64  
 11  CD Account          5000 non-null   int64  
 12  Online              5000 non-null   int64  
 13  CreditCard          5000 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 547.0 KB


Here, there are no null values and no categorical values. No encoding is needed.

In [10]:
UniversalBank.head(20)

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  \
0    1   25           1      49     91107       4    1.6          1         0   
1    2   45          19      34     90089       3    1.5          1         0   
2    3   39          15      11     94720       1    1.0          1         0   
3    4   35           9     100     94112       1    2.7          2         0   
4    5   35           8      45     91330       4    1.0          2         0   
5    6   37          13      29     92121       4    0.4          2       155   
6    7   53          27      72     91711       2    1.5          2         0   
7    8   50          24      22     93943       1    0.3          3         0   
8    9   35          10      81     90089       3    0.6          2       104   
9   10   34           9     180     93023       1    8.9          3         0   
10  11   65          39     105     94710       4    2.4          3         0   
11  12   29           5      45     90277       3    0.1          2         0   
12  13   48          23     114     93106       2    3.8          3         0   
13  14   59          32      40     94920       4    2.5          2         0   
14  15   67          41     112     91741       1    2.0          1         0   
15  16   60          30      22     95054       1    1.5          3         0   
16  17   38          14     130     95010       4    4.7          3       134   
17  18   42          18      81     94305       4    2.4          1         0   
18  19   46          21     193     91604       2    8.1          3         0   
19  20   55          28      21     94720       1    0.5          2         0   

    Personal Loan  Securities Account  CD Account  Online  CreditCard  
0               0                   1           0       0           0  
1               0                   1           0       0           0  
2               0                   0           0       0           0  
3               0                   0           0       0           0  
4               0                   0           0       0           1  
5               0                   0           0       1           0  
6               0                   0           0       1           0  
7               0                   0           0       0           1  
8               0                   0           0       1           0  
9               1                   0           0       0           0  
10              0                   0           0       0           0  
11              0                   0           0       1           0  
12              0                   1           0       0           0  
13              0                   0           0       1           0  
14              0                   1           0       0           0  
15              0                   0           0       1           1  
16              1                   0           0       0           0  
17              0                   0           0       0           0  
18              1                   0           0       0           0  
19              0                   1           0       0           1

In [11]:
UniversalBank.describe()

ID          Age   Experience       Income      ZIP Code  \
count  5000.000000  5000.000000  5000.000000  5000.000000   5000.000000   
mean   2500.500000    45.338400    20.104600    73.774200  93152.503000   
std    1443.520003    11.463166    11.467954    46.033729   2121.852197   
min       1.000000    23.000000    -3.000000     8.000000   9307.000000   
25%    1250.750000    35.000000    10.000000    39.000000  91911.000000   
50%    2500.500000    45.000000    20.000000    64.000000  93437.000000   
75%    3750.250000    55.000000    30.000000    98.000000  94608.000000   
max    5000.000000    67.000000    43.000000   224.000000  96651.000000   

            Family        CCAvg    Education     Mortgage  Personal Loan  \
count  5000.000000  5000.000000  5000.000000  5000.000000    5000.000000   
mean      2.396400     1.937938     1.881000    56.498800       0.096000   
std       1.147663     1.747659     0.839869   101.713802       0.294621   
min       1.000000     0.000000     1.000000     0.000000       0.000000   
25%       1.000000     0.700000     1.000000     0.000000       0.000000   
50%       2.000000     1.500000     2.000000     0.000000       0.000000   
75%       3.000000     2.500000     3.000000   101.000000       0.000000   
max       4.000000    10.000000     3.000000   635.000000       1.000000   

       Securities Account  CD Account       Online   CreditCard  
count         5000.000000  5000.00000  5000.000000  5000.000000  
mean             0.104400     0.06040     0.596800     0.294000  
std              0.305809     0.23825     0.490589     0.455637  
min              0.000000     0.00000     0.000000     0.000000  
25%              0.000000     0.00000     0.000000     0.000000  
50%              0.000000     0.00000     1.000000     0.000000  
75%              0.000000     0.00000     1.000000     1.000000  
max              1.000000     1.00000     1.000000     1.000000

In [12]:
UniversalBank.isna().sum()

ID                    0
Age                   0
Experience            0
Income                0
ZIP Code              0
Family                0
CCAvg                 0
Education             0
Mortgage              0
Personal Loan         0
Securities Account    0
CD Account            0
Online                0
CreditCard            0
dtype: int64

## ENCODING 

In [13]:
# One-hot encoding the family and education columns

family = pd.get_dummies(UniversalBank['Family'], prefix='Family')
education = pd.get_dummies(UniversalBank['Education'], prefix='Education')

# Concatenating one-hot encoded variables with the original data
UniversalBank = pd.concat([UniversalBank,family, education], axis=1)

UniversalBank = UniversalBank.drop(['Family', 'Education'], axis=1)

## 4.0 SPLIT THE DATA

In [14]:
# split the data into validation and training set
train_df, test_df = train_test_split(UniversalBank, test_size=0.3)

# to reduce repetition in later code, create variables to represent the columns
# that are our predictors and target
target = 'CD Account'
predictors = list(UniversalBank.columns)
predictors.remove(target)

## SCALING

In [15]:
# create a standard scaler and fit it to the training set of predictors
scaler = preprocessing.StandardScaler()

# standardizing variables
columns_to_standardize = ["Age", "Experience", "Income", "CCAvg", "Mortgage"]

# Transform the predictors of training and validation sets
train_df[columns_to_standardize] = scaler.fit_transform(train_df[columns_to_standardize]) 

test_df[columns_to_standardize] = scaler.transform(test_df[columns_to_standardize])

## 5.0 SAVE THE DATA

In [16]:
train_X = train_df[predictors]
train_y = train_df[target] # train_target is now a series objecttrain_df.to_csv('RidingMowers_df.csv', index=False)
test_X = test_df[predictors]
test_y = test_df[target] # validation_target is now a series object

## 6.0 MODEL THE DATA

In [17]:
performance = pd.DataFrame({"model": [], "Accuracy": [], "Precision": [], "Recall": [], "F1": []})

## 7.0 LOGISTIC REGRESSION 

## 7.1 LOGISTIC REGRESSION - RANDOM SEARCH

In [18]:
score_measure = "recall"
kfolds = 2

param_grid = {
    'penalty': ['l1','l2','elasticnet'],
    'solver':['saga','liblinear']
}

logistic_reg = LogisticRegression()
rand_search = RandomizedSearchCV(estimator = logistic_reg, param_distributions=param_grid, cv=kfolds, n_iter=10,
                           scoring=score_measure, verbose=1, n_jobs=-1, 
                           return_train_score=True)

_ = rand_search.fit(train_X, train_y)

print(f"The best {score_measure} score is {rand_search.best_score_}")
print(f"... with parameters: {rand_search.best_params_}")

Fitting 2 folds for each of 6 candidates, totalling 12 fits


/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_search.py:305: UserWarning: The total space of parameters 6 is smaller than n_iter=10. Running 6 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


The best recall score is 0.6665971643035863
... with parameters: {'solver': 'liblinear', 'penalty': 'l1'}


/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:378: FitFailedWarning: 
4 fits failed out of a total of 12.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
2 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_logistic.py", line 1291, in fit
    fold_coefs_ = Parallel(n_jobs=self.n_jobs, verbose=self.verbose, prefer=prefer)(
  File "/usr/local/lib/python3.8/dist-packages/sklearn/utils/parallel.py", line 63, in __call__
    return super().__call__(iterable_w

## 7.1 LOGISTIC REGRESSION - GRID SEARCH

In [19]:
score_measure = "recall"
kfolds = 2
penalty = rand_search.best_params_['penalty']
solver = rand_search.best_params_['solver']

param_grid = {  
    'penalty': [penalty],
    'solver': [solver]
}

logistic_reg = LogisticRegression()
grid_search = GridSearchCV(estimator = logistic_reg, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,
                           return_train_score=True)

_ = grid_search.fit(train_X, train_y)

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestRecallLogistic = grid_search.best_estimator_

Fitting 2 folds for each of 1 candidates, totalling 2 fits
The best recall score is 0.6665971643035863
... with parameters: {'penalty': 'l1', 'solver': 'liblinear'}


In [20]:
c_matrix = confusion_matrix(test_y, grid_search.predict(test_X))
TP = c_matrix[1][1]
TN = c_matrix[0][0]
FP = c_matrix[0][1]
FN = c_matrix[1][0]
performance = pd.concat([performance, pd.DataFrame({'model':"Logistic Regression", 
                                                    'Accuracy': [(TP+TN)/(TP+TN+FP+FN)], 
                                                    'Precision': [TP/(TP+FP)], 
                                                    'Recall': [TP/(TP+FN)], 
                                                    'F1': [2*TP/(2*TP+FP+FN)]
                                                     }, index=[0])])

## 8.0 SVM CLASSIFIER 

## 8.1 SVM CLASSIFIER - LINEAR - POLY - RANDOM SEARCH

In [ ]:
score_measure = "recall"
kfolds = 2

param_grid = {
    'C': [10,15],   
    'gamma': ['scale','auto'],
    'kernel':['linear','poly']
}

svm = SVC()
rand_search = RandomizedSearchCV(estimator = svm, param_distributions=param_grid, cv=kfolds, n_iter=15,
                           scoring=score_measure, verbose=1, n_jobs=-1, 
                           return_train_score=True)

_ = rand_search.fit(train_X, train_y)

print(f"The best {score_measure} score is {rand_search.best_score_}")
print(f"... with parameters: {rand_search.best_params_}")

Fitting 2 folds for each of 8 candidates, totalling 16 fits


/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_search.py:305: UserWarning: The total space of parameters 8 is smaller than n_iter=15. Running 8 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


## 8.2 SVM CLASSIFIER - LINEAR - POLY - GRID SEARCH

In [ ]:
score_measure = "recall"
kfolds = 2

C = rand_search.best_params_['C']
gamma = rand_search.best_params_['gamma']
kernel = rand_search.best_params_['kernel']

param_grid = {
    'C': np.arange(C-1,C+1),  
    'gamma': [gamma],
    'kernel': [kernel]
    
}

svm = SVC()
grid_search = GridSearchCV(estimator = svm, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,
                           return_train_score=True)

_ = grid_search.fit(train_X, train_y)

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestRecallSVM = grid_search.best_estimator_

In [ ]:
c_matrix = confusion_matrix(y_test, grid_search.predict(X_test))
TP = c_matrix[1][1]
TN = c_matrix[0][0]
FP = c_matrix[0][1]
FN = c_matrix[1][0]
performance = pd.concat([performance, pd.DataFrame({'model':"SVM", 
                                                    'Accuracy': [(TP+TN)/(TP+TN+FP+FN)], 
                                                    'Precision': [TP/(TP+FP)], 
                                                    'Recall': [TP/(TP+FN)], 
                                                    'F1': [2*TP/(2*TP+FP+FN)]
                                                     }, index=[0])])

## 9.0 DECISION TREE

## 9.1 DECISION TREE - RANDOM SEARCH 

In [24]:
score_measure = "recall"
kfolds = 2

param_grid = {
    'min_samples_split': np.arange(2,50),  
    'min_samples_leaf': np.arange(1,50),
    'min_impurity_decrease': np.arange(0.0001, 0.01, 0.0005),
    'max_leaf_nodes': np.arange(5, 50), 
    'max_depth': np.arange(1,50), 
    'criterion': ['entropy', 'gini'],
}

dtree = DecisionTreeClassifier()
rand_search = RandomizedSearchCV(estimator = dtree, param_distributions=param_grid, cv=kfolds, n_iter=50,
                           scoring=score_measure, verbose=1, n_jobs=-1, 
                           return_train_score=True)

_ = rand_search.fit(train_X, train_y)

print(f"The best {score_measure} score is {rand_search.best_score_}")
print(f"... with parameters: {rand_search.best_params_}")

Fitting 2 folds for each of 50 candidates, totalling 100 fits
The best recall score is 0.6849040867389491
... with parameters: {'min_samples_split': 46, 'min_samples_leaf': 1, 'min_impurity_decrease': 0.0081, 'max_leaf_nodes': 29, 'max_depth': 29, 'criterion': 'entropy'}


## 9.1 DECISION TREE - GRID SEARCH

In [26]:
score_measure = "recall"
kfolds = 2
min_samples_split = rand_search.best_params_['min_samples_split']
min_samples_leaf = rand_search.best_params_['min_samples_leaf']
min_impurity_decrease = rand_search.best_params_['min_impurity_decrease']
max_leaf_nodes = rand_search.best_params_['max_leaf_nodes']
max_depth = rand_search.best_params_['max_depth']
criterion = rand_search.best_params_['criterion']

param_grid = {
    'min_samples_split': np.arange(min_samples_split-2,min_samples_split+2),  
    'min_samples_leaf': np.arange(min_samples_leaf-2,min_samples_leaf+2),
    'min_impurity_decrease': np.arange(min_impurity_decrease-0.0001, min_impurity_decrease+0.0001, 0.00005),
    'max_leaf_nodes': np.arange(max_leaf_nodes-2,max_leaf_nodes+2), 
    'max_depth': np.arange(max_depth-2,max_depth+2), 
    'criterion': [criterion]
}

dtree = DecisionTreeClassifier()
grid_search = GridSearchCV(estimator = dtree, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,
                           return_train_score=True)

_ = grid_search.fit(train_X, train_y)

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestRecallTree = grid_search.best_estimator_

Fitting 2 folds for each of 1024 candidates, totalling 2048 fits
The best recall score is 0.6849040867389491
... with parameters: {'criterion': 'entropy', 'max_depth': 27, 'max_leaf_nodes': 27, 'min_impurity_decrease': 0.008, 'min_samples_leaf': 1, 'min_samples_split': 46}


/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:378: FitFailedWarning: 
1024 fits failed out of a total of 2048.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
512 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/usr/local/lib/python3.8/dist-packages/sklearn/tree/_classes.py", line 889, in fit
    super().fit(
  File "/usr/local/lib/python3.8/dist-packages/sklearn/tree/_classes.py", line 177, in fit
    self._validate_params()
  File "/usr/local/lib/python3.8/dist-packages/sklearn/base.py", line 581, in _validat

In [28]:
c_matrix = confusion_matrix(test_y, grid_search.predict(test_X))
TP = c_matrix[1][1]
TN = c_matrix[0][0]
FP = c_matrix[0][1]
FN = c_matrix[1][0]
performance = pd.concat([performance, pd.DataFrame({'model':"Decision Tree", 
                                                    'Accuracy': [(TP+TN)/(TP+TN+FP+FN)], 
                                                    'Precision': [TP/(TP+FP)], 
                                                    'Recall': [TP/(TP+FN)], 
                                                    'F1': [2*TP/(2*TP+FP+FN)]
                                                     }, index=[0])])

## PERFORMANCE OF ALL MODELS

In [29]:
performance.sort_values(by=['Recall'])

model  Accuracy  Precision    Recall        F1
0  Logistic Regression     0.978   1.000000  0.602410  0.751880
0        Decision Tree     0.976   0.912281  0.626506  0.742857

## ANALYSIS OF MODELS 

We considered three kind of models here for the CD Account holding. The Recall score where the FN(False Negative) is more prioritized here compared to the FP(False positive). The Recall score of Logistic Regression model is 0.602410, and  Decision tree model is 0.626506. As we are considering the Recall score as priority, Decision tree with 0.626506 as recall score is better model compared to other model.